src/client/replicon_client.rs

```diff
@@ -0,0 +1,184 @@
+            .received_messages
+            .get_mut(channel_id as usize)
+            .unwrap_or_else(|| panic!("client should have a receive channel with id {channel_id}"));
+
+        channel_messages.pop()
+    }
+
+    /// Sends a message to the server over a channel.
+    pub fn send<I: Into<u8>, B: Into<Bytes>>(&mut self, channel_id: I, message: B) {
+        if !self.is_connected() {
+            warn!("trying to send a message when the client is not connected");
+            return;
+        }
+
+        let channel_id = channel_id.into();
+        let channel_messages = self
+            .sent_messages
+            .get_mut(channel_id as usize)
+            .unwrap_or_else(|| panic!("client should have a send channel with id {channel_id}"));
+
+        channel_messages.push(message.into());
+    }
+
+    /// Sets the client connection status.
+    ///
+    /// Should be called only from the messaging backend when the client status changes.
+    /// Discards all messages if the state changes from [`RepliconClientStatus::Connected`].
+    /// See also [`Self::status`].
+    pub fn set_status(&mut self, status: RepliconClientStatus) {
+        if self.is_connected() && !matches!(status, RepliconClientStatus::Connected { .. }) {
+            for channel_messages in &mut self.sent_messages {
+                channel_messages.clear();
+            }
+            for channel_messages in &mut self.received_messages {
+                channel_messages.clear();
+            }
+        }
+
+        self.status = status;
+    }
+
+    /// Returns the current client status.
+    ///
+    /// See also [`Self::set_status`].
+    #[inline]
+    pub fn status(&self) -> RepliconClientStatus {
+        self.status
+    }
+
+    /// Returns `true` if the client is disconnected.
+    ///
+    /// See also [`Self::status`].
+    #[inline]
+    pub fn is_disconnected(&self) -> bool {
+        matches!(self.status, RepliconClientStatus::Disconnected)
+    }
+
+    /// Returns `true` if the client is connecting.
+    ///
+    /// See also [`Self::status`].
+    #[inline]
+    pub fn is_connecting(&self) -> bool {
+        matches!(self.status, RepliconClientStatus::Connecting)
+    }
+
+    /// Returns `true` if the client is connected.
+    ///
+    /// See also [`Self::status`].
+    #[inline]
+    pub fn is_connected(&self) -> bool {
+        matches!(self.status, RepliconClientStatus::Connected { .. })
+    }
+
+    /// Returns the client's ID.
+    ///
+    /// The client ID is available only if the client state is [`RepliconClientStatus::Connected`].
+    /// See also [`Self::status`].
+    #[inline]
+    pub fn id(&self) -> Option<ClientId> {
+        if let RepliconClientStatus::Connected { client_id } = self.status {
+            client_id
+        } else {
+            None
+        }
+    }
+
+    /// Returns an iterator over all messages for each channel.
+    ///
+    /// Should be called only by the messaging backend.
```

Comment:
Yes, draining was my first attempt, but I had problems using it with multiple clients in benchmarks...

But I think I can solve it nicely now since I changed how tests works. Let me try.